[CALCITE-7112] Eliminate nested calls for idempotent unary functions upper/lower/abs

xuzifu666 opened the pull request:
jira: https://issues.apache.org/jira/browse/CALCITE-7122

xiedeyantu commented on core/src/main/java/org/apache/calcite/rex/RexSimplify.java:
```diff
@@ -397,6 +403,22 @@ RexNode isFalse(RexNode e) {
+  /**
+   * Runs simplification unary function by eliminating idempotent.
+   */
+  private RexCall simplifUnaryFunction(RexCall rexCall) {
```
Maybe we can simplify `lower(upper(lower(x)))` to `upper(lower(x))` or `lower(upper(x))`?

xuzifu666 replied:
This is a bit like a special case, because the two functions have opposite effects. Maybe can hand this case, and it will not affect the default behavior.

xuzifu666 replied:
Yes，currently this goal can be achieved based on this PR. For example, lower(lower(upper(uper('a')))) will be simplified to lower(upper('a')).

xuzifu666 replied:
As @julianhyde  provided in the case of Jira, these two functions do not have opposite effects, so this case is not need to be simplified.

xuzifu666 commented on core/src/main/java/org/apache/calcite/rex/RexSimplify.java:
```diff
@@ -333,7 +337,9 @@ RexNode simplify(RexNode e, RexUnknownAs unknownAs) {
       if (e.getClass() == RexCall.class) {
-        return simplifyGenericNode((RexCall) e);
+        RexCall rexCall = (RexCall) e;
+        rexCall = simplifUnaryFunction(rexCall);
```
I was going to do this at first, but it is handled according to SqlKind. The default SqlKind of upper/lower/abs is OTHER_FUNCTION. If I need to add a new SqlKind and modify the SqlKind of these functions because of this, it doesn’t feel necessary.

xuzifu666 commented on core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java:
```diff
@@ -361,6 +366,59 @@ public class RexImplicationCheckerTest {
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-7122">[CALCITE-7122]
+   * Eliminate nested calls for idempotent unary functions UPPER/LOWER/ABS/INITCAP</a>. */
+  @Test void testSimplifyIdempotentUnaryFunctions() {
```
In order to facilitate intuitive understanding of the case, I added SQL testing.

julianhyde replied:
No. `RexImplicationCheckerTest` is a unit test. If you drag in the functionality that goes to and from SQL, it is not longer a unit test.

To make the tests easier to understand, add comments. E.g. "Test that abs(abs(x)) is simplified to abs(x)".

xuzifu666 replied:
OK, I would add more detail comments for these tests and remove sql tests. @julianhyde

ILuffZhe reviewed (approved):
LGTM! Let's see if there are new comments come in the next 2 days.

xiedeyantu commented:
Can idempotent functions such as ceil and floor also be considered together?

xuzifu666 commented:
> Can idempotent functions such as ceil and floor also be considered together?

The ceil and floor functions have already been separately processed before, and the logic is a little complex, so they should be no need to handle them here.

julianhyde commented:
> I see, these two functions are binary functions. 

Numeric `CEIL` and `FLOOR` are unary. E.g. `CEIL(CEIL(3.14))` = `CEIL(3.14)` = `4`.

Datetime `CEIL` and `FLOOR` are binary. E.g. `FLOOR(FLOOR(DATE '2025-02-08' TO MONTH))` = `FLOOR(DATE '2025-02-08' TO MONTH)` = `DATE '2025-02-01'`. The second parameter is a time unit, basically a literal over an enum type, so we could regard the functions as a family of unary functions, e.g. `FLOOR_MONTH`, `CEIL_MONTH`, `FLOOR_YEAR`, etc.

xiedeyantu commented:
> > I see, these two functions are binary functions.
> 
> Numeric `CEIL` and `FLOOR` are unary. E.g. `CEIL(CEIL(3.14))` = `CEIL(3.14)` = `4`.
> 
> Datetime `CEIL` and `FLOOR` are binary. E.g. `FLOOR(FLOOR(DATE '2025-02-08' TO MONTH))` = `FLOOR(DATE '2025-02-08' TO MONTH)` = `DATE '2025-02-01'`. The second parameter is a time unit, basically a literal over an enum type, so we could regard the functions as a family of unary functions, e.g. `FLOOR_MONTH`, `CEIL_MONTH`, `FLOOR_YEAR`, etc.

I agree with @julianhyde 's idea. I also have a suggestion—though I’m not sure how feasible it is—to consolidate all the currently identified unary idempotent functions into a unified encapsulation while ensuring compatibility with the existing ones in  `simplify` . The goal would be to extract a shared logic layer to avoid scattering similar implementations across multiple places, which could increase future maintenance costs. Currently, only four functions are handled uniformly, and without careful review, others might mistakenly treat  CEIL  and  FLOOR  as non-idempotent or assume they were overlooked. If modifying this proves too difficult, we could at least document all the relevant functions in the comments to prevent future developers from thinking the feature was implemented incompletely.
I’m not entirely sure if this suggestion is valid. If Julian has time, I’d appreciate it if you could take a look as well.

mihaibudiu commented:
what is the status of this PR? It is marked as "changes requested". Have all change request been addressed?

xuzifu666 commented:
> what is the status of this PR? It is marked as "changes requested". Have all change request been addressed?

IMO it had been addressed all comments now. @mihaibudiu

mihaibudiu commented:
So your suggestion is to add a property to SqlFunction which says whether the function is idempotent?
Frankly, I think there would be too many such properties to create for various interesting semantic properties of functions that could be used by the optimizer.
Moreover, I think that the kinds of programs optimized by this PR never occur in practice, because this is just not how people write programs. So I think this is a lot of work for very little benefit.
